### DISCUSSION TITLE
Putting null keys on a `MapMessage` shouldn't be throwing exception

### DISCUSSION THREAD

**Discussion by SubieGold** [apache/logging-log4j2#2009] [General]
Looks like using MapMessage with null values causes a IllegalArgumentException to be thrown. I think that MapMessage should be able to accept nulls or the logger should be able to handle the exception.

https://github.com/apache/logging-log4j2/blob/613b520f650af4dbc74d42f62b9931bfb71569d4/log4j-api/src/main/java/org/apache/logging/log4j/message/MapMessage.java#L205

**Comment by vy** [comment]
@LoganDev99, you might indeed be right. Do you have a particular use case that caused this exception?

**Reply by SubieGold** [reply]
We were logging some data from another API that unfortunately returned a null. Logging the null would have been perfectly fine since the application would have just continued.

**Reply by vy** [reply]
Okay, I see your point. I don't know whether `MapMessage` (and its subclasses) should register an entry with `null` key or simply discard it. I would like to hear your thoughts on this too.

All in all, feel free to submit a PR. Though note that I would prefer to see this new behaviour in `MapMessage` subclasses too.

**Reply by SubieGold** [reply]
Perhaps it can accept the key and value and just log string null?

**Reply by ppkarwasz** [reply]
> Okay, I see your point. I don't know whether `MapMessage` (and its subclasses) should register an entry with `null` key or simply discard it. I would like to hear your thoughts on this too.

For reference SLF4J discards all calls to `MDC.pushByKey` if the key is `null` (its implementation is based on [`ThreadLocalMapOfStacks`](https://github.com/qos-ch/slf4j/blob/master/slf4j-api/src/main/java/org/slf4j/helpers/ThreadLocalMapOfStacks.java)).

I think that discarding it is the safest choice: if the context data is copied into a map that does not accept `null`s it could cause problems downstream.

**Comment by rgoers** [comment]
@ppkarwasz Comparing how SLF4J handles a ThreadContext map entry vs how we handle a MapMessage should not be considered to have the same behavior. That said, I don't believe we should always accept nulls. At best, it is fair to say that the behavior when a null value is passed is undefined (or up to the implementation).

**Reply by ppkarwasz** [reply]
Agreed,

I was a little bit bluffing with the `null` key. As far as our `MDCAdapter` implementation goes, I use a `null` key passed to `pushByKey/popByKey` to allow users to access our NDC.
